sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,777 @@
+    # as local maxima points
+    for i in range(1, len(reachability_ordering) - 1):
+        # if the point is a local maxima on the reachability plot with
+        # regard to neighborhood_size, insert it into priority queue and
+        # maxima list
+        if (reachability_plot[i] > reachability_plot[i - 1] and
+            reachability_plot[i] >= reachability_plot[i + 1] and
+            _is_local_maxima(i, reachability_plot, reachability_ordering,
+                             neighborhood_size) == 1):
+            local_maxima_points[i] = reachability_plot[i]
+
+    return sorted(local_maxima_points,
+                  key=local_maxima_points.__getitem__, reverse=True)
+
+
+def _cluster_tree(node, parent_node, local_maxima_points,
+                  reachability_plot, reachability_ordering,
+                  min_cluster_size, maxima_ratio, rejection_ratio,
+                  similarity_threshold, significant_min):
+    """Recursively builds cluster tree to hold hierarchical cluster structure
+
+    node is a node or the root of the tree in the first call
+    parent_node is parent node of N or None if node is root of the tree
+    local_maxima_points is list of local maxima points sorted in
+    descending order of reachability
+    """
+
+    if len(local_maxima_points) == 0:
+        return  # parent_node is a leaf
+
+    # take largest local maximum as possible separation between clusters
+    s = local_maxima_points[0]
+    node.assign_split_point(s)
+    local_maxima_points = local_maxima_points[1:]
+
+    # create two new nodes and add to list of nodes
+    node_1 = _TreeNode(reachability_ordering[node.start:s],
+                       node.start, s, node)
+    node_2 = _TreeNode(reachability_ordering[s + 1:node.end],
+                       s + 1, node.end, node)
+    local_max_1 = []
+    local_max_2 = []
+
+    for i in local_maxima_points:
+        if i < s:
+            local_max_1.append(i)
+        if i > s:
+            local_max_2.append(i)
+
+    node_list = []
+    node_list.append((node_1, local_max_1))
+    node_list.append((node_2, local_max_2))
+
+    if reachability_plot[s] < significant_min:
+        node.assign_split_point(-1)
+        # if split_point is not significant, ignore this split and continue
+        _cluster_tree(node, parent_node, local_maxima_points,
+                      reachability_plot, reachability_ordering,
+                      min_cluster_size, maxima_ratio, rejection_ratio,
+                      similarity_threshold, significant_min)
+        return
+
+    # only check a certain ratio of points in the child
+    # nodes formed to the left and right of the maxima
+    # ...should check_ratio be a user settable parameter?
+    check_ratio = .8
+    check_value_1 = int(np.round(check_ratio * len(node_1.points)))
+    check_value_2 = int(np.round(check_ratio * len(node_2.points)))
+    if check_value_2 == 0:
+        check_value_2 = 1
```

Comment:
I'm commenting out this conditional statement for now.

I can't think of (or code) any possible case when it would occur; it's rounding 0.8 * len(node_2), then checking if it was rounded down to zero. The len of the points list will aways be an integer, and the only valid integer that rounds down to 0 is 0.8 times 0. node_2 point list is determined by taking a slice from a local max forward to some number of reachability entries... but the local_max function can't assign the last reachability distance as a local max, as it determines local maxes by looking at the left and right numbers in the array-- i.e., you can flag the second to last reachability distance as a local max, but not the last entry. So the smallest slice possible for node2 points is 1 + the second to last entry in the reachability graph to the end of the array... which will always be length of 1 and rounded to 1.